initial schema and database configuration

jmartin-sul opened the pull request:
A first pass at the tables and models needed for the object inventory store: preserved_objects (a master record covering all preserved copies for a given druid), endpoints (unique replication targets), and preservation_copies (individual preserved copies of preservation objects, on specific endpoints).

paired w/ @SaravShah on implementation.  met w/ @LynnMcRae and @julianmorley to refine schema choices (meeting w/ @julianmorley again this afternoon for further schema refinement).

supersedes #28 

Resolves #16
Resolves #17
Resolves #31

ndushay commented on app/views/moab_storage/show.html.erb:
```diff
@@ -0,0 +1 @@
+<h1> Test </h1>
```
This shouldn't be here -- we have no views.  If someone wants this locally, they can add it to .gitignore

jmartin-sul replied:
yeah, not sure how this got added, will remove.

tingulfsen commented on spec/models/endpoint_spec.rb:
```diff
@@ -0,0 +1,24 @@
+
+  it 'enforces unique constraint on endpoint_name' do
+    expect do
+      Endpoint.create!(endpoint_name: 'aws', endpoint_type: 'cloud')
```
How about changing the endpoint_type to be something other than 'cloud' just to emphasize the 'endpoint_name' requirement?

jmartin-sul replied:
i think we're already going to revisit this (per #46) and turn `endpoint_type` into an FK to a table that has richer endpoint info.  so i'm inclined to leave this as-is, just to not tweak something that'll be changing in a different way anyway.

tingulfsen replied:
👍

tingulfsen commented on spec/models/endpoint_spec.rb:
```diff
@@ -0,0 +1,24 @@
+
+  let!(:endpoint) { Endpoint.create(endpoint_name: 'aws', endpoint_type: 'cloud') }
+
+  it 'is not valid without valid attributes' do
```
Can we add another test where just 1 of the 2 requirements are fulfilled? Or am I too anal?

ndushay replied:
Did this get addressed by added specs below?

tingulfsen commented on app/models/endpoint.rb:
```diff
@@ -0,0 +1,8 @@
+class Endpoint < ApplicationRecord
+  has_many :preservation_copies
+  validates :endpoint_name, presence: true, uniqueness: true
+  validates :endpoint_type, presence: true
```
Do we expect 'endpoint_type' to be a finite set and if so does it make sense to represent it as such?

jmartin-sul replied:
i think we do, but per #46, there will be a table containing a list of endpoint types.  otherwise, i'd agree that we should constrain the value to known good types.

tingulfsen commented on spec/models/preservation_copy_spec.rb:
```diff
@@ -0,0 +1,27 @@
+
+RSpec.describe PreservationCopy, type: :model do
+  let!(:preservation_object2) do
+    PreservedObject.create!(druid: 'ab123cd45679', version: 1, preservation_policy: 'keepit', size: 1)
```
Do we expect that 'preservation_policy' will be a small finite set, and if so does it make sense to represent it as such?

jmartin-sul replied:
similar to endpoint type, there'll be a policy definition table that this will point to (see #46).  so i'm inclined to leave this as-is and revisit it when we work on #46 (which should happen either later this week, or first thing next week).

tingulfsen replied:
👍

tingulfsen commented on db/migrate/20170823232731_create_preserved_objects.rb:
```diff
@@ -0,0 +1,12 @@
+  def change
+    create_table :preserved_objects do |t|
+      t.string :druid, null: false, unique: true, index: true
+      t.integer :version
```
When would we expect 'version' to be null or not present? Should this also be a case of 'null: false'?

jmartin-sul replied:
i'm not sure... i had a vague thought along these lines when we were first defining the schema last week.  but what, e.g., if you had a corrupted moab where you couldn't determine the version?  in that case, i think we'd want a catalog entry, but i don't know that we could put a valid integer in this field.  though this is not a use case i'm sure about at all.

julianmorley replied:
What would a corrupted Moab look like? Presumably missing version directories, or missing expected files inside one or more version directories. Would we ever even ingest such a Moab into OIS, or violently throw up before we even got to that point? If a known-good, previously-sane Moab later has problems (v7 is fine, but v8 is added and is bad), would we record that Moab as being v7, but with problems, or set version to 'null' and and then handle the exception somehow?

julianmorley replied:
That's a round-about way of saying I think requiring a value for version is probably desirable, at the very least as a way of tracking "the last known-good version" of a Moab if we run into subsequent problems.

jmartin-sul replied:
i guess i was imagining something like... we create a top level moab directory for an object, and something dies silently before the first version subdirectory is created.  other processes expect a valid moab object to be there, but while there was a place made for it, a valid moab object was never fully created in the first place.

if this is completely unrealistic, and no similar situation can be imagined, i'd agree that this field should be required.

julianmorley replied:
I think in that scenario, either the inventory process or ingest should throw up long before we try to write data to the database. We should take it as a point of pride that we don't ingest bad data and alert very noisily if that happens.

There's also the side-question of "what makes a directory a Moab object?"; we shouldn't pin that to 'the directory is in Moab name format', but rather that the contents of that directory contain everything we'd expect to see in a well-formed Moab, and nothing else.

LynnMcRae replied:
Agreed. We don't have a scenario where we're building up a Moab object from an initially empty directory. It only exists to contain at least one version of the object, so anything short of that should fail to be created.

jmartin-sul replied:
thanks for the info, sounds good, will add the constraint

tingulfsen commented on spec/models/preservation_copy_spec.rb:
```diff
@@ -0,0 +1,27 @@
+
+  it 'is not valid without valid attributes' do
+    expect(PreservationCopy.new).not_to be_valid
+  end
```
Same as per my comment above - is there a sane and non-anal way to test that the correct attributes are present without generating all possible permutations?

jmartin-sul commented on .travis.yml:
```diff
@@ -1,4 +1,12 @@
+  - psql -c 'ALTER USER preservation_core_catalog CREATEDB;' -U postgres
+  - psql -c 'CREATE DATABASE preservation_core_catalog;' -U postgres
+  - psql -c 'ALTER DATABASE preservation_core_catalog OWNER TO preservation_core_catalog;' -U postgres
+  - psql -c 'GRANT ALL PRIVILEGES ON DATABASE preservation_core_catalog TO preservation_core_catalog;' -U postgres
```
yeah, a script would be reasonable.  see #48.

jmartin-sul commented on Gemfile:
```diff
@@ -21,12 +19,19 @@ gem 'jbuilder', '~> 2.5'
+
+# useful for debugging, even in prod
+gem 'pry-byebug' # Adds step-by-step debugging and stack navigation capabilities to pry using byebug
+gem 'pry-rails' # use pry as the rails console shell instead of IRB
```
i've definitely used rails console a number of times on prod to investigate argo issues.  not a thing that i'd expect to do often, but something i'd expect to come up at least some.  e.g., you want to run load up some objects via ActiveRecord to because some error cropped up when a thing you expected to be there wasn't.  and in that case, the syntax highlighting and formatting and inter-session history and other niceties of pry are great.

jmartin-sul commented on README.md:
````diff
@@ -45,20 +45,23 @@ If you use homebrew you can install PostgreSQL by typing  `brew install postgres
+CREATE DATABASE preservation_core_catalog;
+ALTER DATABASE preservation_core_catalog OWNER TO preservation_core_catalog;
+GRANT ALL PRIVILEGES ON DATABASE preservation_core_catalog TO preservation_core_catalog;
+```
````
could probably script this, let's say tentatively see #48 for this one too.

jmartin-sul commented:
@ndushay:  i think i've made the requested fixes (i added a commit that just generally cleaned up Gemfile so that its layout is organized more like our other projects).  for the things that you said we could ticket, i created a ticket.

@tingulfsen:  will discuss the testing suggestions w/ @SaravShah

jmartin-sul commented:
@tingulfsen:  we added an invalid permutation to the tests for each of the things you mentioned.  we didn't do all possible permutations, partially because that part of the schema is going to change, and partially because i don't think we typically do every possible permutation for that sort of thing?  i know there are cases where we've done tons of permutations (e.g. the cancan tests that we added to stacks during the recent IIIF work), but i think we typically just do a few representative cases, right?

@LynnMcRae and @julianmorley:  added the not null constraint to the version column (and added a corresponding test similar to the ones @tingulfsen suggested above).

ndushay commented on Gemfile:
```diff
@@ -5,28 +5,28 @@ git_source(:github) do |repo_name|
+gem 'rails', '~> 5.1.3'
 
-gem 'config'
+# Stanford/Samvera gems
```
Samvera???   ew.
